site/src/pages/TemplateSettingsPage/TemplateVariablesPage/TemplateVariablesPage.tsx

```diff
@@ -6,40 +6,103 @@ import {
-import { FC } from "react";
+import { useEffect, type FC, useState } from "react";
 import { Helmet } from "react-helmet-async";
 import { useNavigate, useParams } from "react-router-dom";
-import { templateVariablesMachine } from "xServices/template/templateVariablesXService";
-import { pageTitle } from "../../../utils/page";
+import { pageTitle } from "utils/page";
 import { useTemplateSettings } from "../TemplateSettingsLayout";
 import { TemplateVariablesPageView } from "./TemplateVariablesPageView";
+import {
+  createTemplateVersion,
+  templateVersion,
+  templateVersionVariables,
+  updateActiveTemplateVersion,
+} from "api/queries/templates";
+import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
+import { ErrorAlert } from "components/Alert/ErrorAlert";
+import { Loader } from "components/Loader/Loader";
 
 export const TemplateVariablesPage: FC = () => {
   const { template: templateName } = useParams() as {
     organization: string;
     template: string;
   };
-  const organizationId = useOrganizationId();
+  const orgId = useOrganizationId();
   const { template } = useTemplateSettings();
   const navigate = useNavigate();
-  const [state, send] = useMachine(templateVariablesMachine, {
-    context: {
-      organizationId,
-      template,
-    },
-    actions: {
-      onUpdateTemplate: () => {
-        displaySuccess("Template updated successfully");
+  const queryClient = useQueryClient();
+  const versionId = template.active_version_id;
+
+  const [newVersionId, setNewVersionId] = useState<string | null>(null);
+
+  const {
+    data: version,
+    error: versionError,
+    isLoading: isVersionLoading,
+  } = useQuery({ ...templateVersion(versionId), keepPreviousData: true });
+  const {
+    data: variables,
+    error: variablesError,
+    isLoading: isVariablesLoading,
+  } = useQuery({
+    ...templateVersionVariables(versionId),
+    keepPreviousData: true,
+  });
+
+  const createVersionMutation = createTemplateVersion(orgId, queryClient);
+  const { mutate: sendCreateTemplateVersion, error: createVersionError } =
+    useMutation({
+      ...createVersionMutation,
+      onSuccess: (data) => {
+        setNewVersionId(data.id);
       },
+    });
+  const publishMutation = updateActiveTemplateVersion(template, queryClient);
+  const { mutate: publishVersion, error: publishError } = useMutation({
+    ...publishMutation,
+    onSuccess: () => {
+      publishMutation.onSuccess();
+      displaySuccess("Template updated successfully");
+    },
+    onSettled: () => {
+      setNewVersionId(null);
     },
   });
+
   const {
-    activeTemplateVersion,
-    templateVariables,
-    getTemplateDataError,
-    updateTemplateError,
-    jobError,
-  } = state.context;
+    data: status,
+    error: statusError,
+    refetch: refetchStatus,
+    isRefetching,
+  } = useQuery(
+    newVersionId ? templateVersion(newVersionId) : { enabled: false },
+  );
```

Comment:
I'm wondering why not have a single function to wrap the publishing logic then split that into many queries 🤔 :

Pseudo code:
```ts
function publishNewTemplateVersion(variables: Variables) {
  await createNewTemplateVersion(variables)

  let isBuilding = true
  while (isBuilding) {
    const buildStatus = await getBuildStatus(variables)
    isBuilding = buildStatus === 'building'
    if(isBuilding) {
      await sleep(2_000)
    } else {
      isBuilding = false
    }
  }
}
```
The watching part can also be extracted into a generator function `watchBuild` or something like that if it makes sense. I just think it is better than gluing multiple queries and mutations together for one single operation. Makes sense?